src/bank.rs

```diff
@@ -793,23 +903,28 @@ mod tests {
         // Assert the debit counts as a transaction.
         assert_eq!(bank.transaction_count(), 1);
 
-        // Mint's balance will be zero because all funds are locked up.
-        assert_eq!(bank.get_balance(&mint.pubkey()), 0);
+        // Mint's balance will be 1 because 1 of the tokens is locked up.
+        assert_eq!(bank.get_balance(&mint.pubkey()), 1);
 
-        // pubkey's balance will be None because the funds have not been
-        // sent.
+        // pubkey's balance will be 0 because the funds are locked up
         assert_eq!(bank.get_balance(&pubkey), 0);
 
         // Now, cancel the trancaction. Mint gets her funds back, pubkey never sees them.
-        bank.apply_signature(mint.pubkey(), signature).unwrap();
-        assert_eq!(bank.get_balance(&mint.pubkey()), 1);
+        let tx = Transaction::new_signature(&mint.keypair(), pubkey, signature, bank.last_id());
+        let res = bank.process_transaction(&tx);
+        assert!(res.is_ok());
         assert_eq!(bank.get_balance(&pubkey), 0);
+        assert_eq!(bank.get_balance(&mint.pubkey()), 2);
 
-        // Assert cancel doesn't cause count to go backward.
-        assert_eq!(bank.transaction_count(), 1);
+        // Assert cancel counts as a tx
+        assert_eq!(bank.transaction_count(), 2);
 
-        bank.apply_signature(mint.pubkey(), signature).unwrap(); // <-- Attack! Attempt to cancel completed transaction.
-        assert_ne!(bank.get_balance(&mint.pubkey()), 2);
+        //try to replay the signature contract
+        bank.register_entry_id(&hash(bank.last_id().as_ref()));
+        let tx = Transaction::new_signature(&mint.keypair(), pubkey, signature, bank.last_id());
+        let res = bank.process_transaction(&tx);
```

Comment:
Why not use `apply_signature`?  And why delete that `Attack!` comment?

Reply:
@garious `apply_signature` used to update an internal bank state directly.  so it bypassed all the transaction plumbing to get to the correct context.  budget dsl needs to be pulled out of bank, and have its own internal unit tests that work on the state transitions without the bank's plumbing involved.